### PR TITLE
Add pbr for v1

### PR DESCRIPTION
**Pull request by Joseph-Edwards**
This updates pbr in preparation for v1. Since, in `libsemigroups` there is no "checks" version of multiplying `PBR`s, you can't really do very much with these objects, but they're here nonetheless!

### CODE REVIEW

**Comment by Joseph-Edwards** [comment]
The doc will have some bad signatures until #190 is merged.

File: src/pbr.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+:raises LibsemigroupsError: if the resultant PBR:
+
+    *  would not describe a binary relation on an even number of points; or
+    *  would have a point related to a point that is greater than degree();
+)pbdoc");
+    thing.def(py::init([](PBR::vector_type<uint32_t> x) { return to_pbr(x); }),
+              py::arg("x"),
+              R"pbdoc(
+Construct from adjacencies 0 to :math:`2n - 1`.
```
Comment (james-d-mitchell) [inline src/pbr.cpp:138]: 0 should be inside double backticks here

File: src/pbr.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+and the vector in position :math:`i` is the list of points adjacent to :math:`i`
+in the :any:`PBR` constructed.
+
+:param x: the container of vectors of adjacencies.
+:type x: List[List[int]]
+
+:raises LibsemigroupsError: if the resultant PBR:
+
+    *  would not describe a binary relation on an even number of points; or
+    *  would have a point related to a point that is greater than degree();
```
Comment (james-d-mitchell) [inline src/pbr.cpp:151]: Reference degree?

File: src/pbr.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+Returns the number of points of a :any:`PBR`.
+
+:returns: A value of type ``int``.
+:rtype: int
+
+:complexity: Constant.
+)pbdoc");
+    thing.def(
+        "__getitem__",
+        [](const PBR& a, size_t b) { return a[b]; },
```
Comment (james-d-mitchell) [inline src/pbr.cpp:181]: Is this safe if b is out of bounds?

File: src/pbr.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        py::arg("x"),
+        R"pbdoc(
+Returns the identity PBR with degree ``x.degree()``.
+
+This member function returns a new :any:`PBR` with degree equal to the degree of
+*x* , where every value is adjacent to its negative. Equivalently, :math:`i` is
+adjacent :math:`i + n` and vice versa for every :math:`i` less than the degree
+:math:`n`.
+
+:param x: The identity.
```
Comment (james-d-mitchell) [inline src/pbr.cpp:208]: This seems sus

File: tests/test_pbr.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+"""
+This module contains some tests for PBRs.
+"""
+
+# pylint: disable= missing-function-docstring, invalid-name
+
+from copy import copy
+import pytest
+from libsemigroups_pybind11 import PBR, pbr
+
```
Comment (james-d-mitchell) [inline tests/test_pbr.py:18]: Maybe add some me tests for the left right constructor?

File: src/pbr.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    *  would have a point related to a point that is greater than degree();
+    *  *x* contains a list of points related to a point that is not sorted.
+
+)pbdoc");
+    thing.def("degree",
+              &PBR::degree,
+              R"pbdoc(
+Returns the degree of a :any:`PBR`.
+
+Returns the degree of a :any:`PBR` , where the *degree* of a :any:`PBR` is half
```
Comment (james-d-mitchell) [inline src/pbr.cpp:160]: Extra space before comma

File: src/pbr.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+:type n: int
+)pbdoc");
+    thing.def(py::init([](PBR::vector_type<int32_t> left,
+                          PBR::vector_type<int32_t> right) {
+                return to_pbr(left, right);
+              }),
+              py::arg("left"),
+              py::arg("right"),
+              R"pbdoc(
+Construct from adjacencies 1 to :math:`n` and -1 to :math:`-n`.
```
Comment (james-d-mitchell) [inline src/pbr.cpp:115]: Mixture of math, plain numbers and double back ticks

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]

**Comment by james-d-mitchell** [comment]
Can you maybe add the checked multiplication, and the other comparison operators? Shouldn't be much code

File: src/pbr.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    thing.def(
+        "copy",
+        [](PBR const& self) { return PBR(self); },
+        R"pbdoc(
+Copy a PBR object.
+
+:returns: A copy.
+:rtype: PBR
+)pbdoc");
+    thing.def(py::init<size_t>(),
```
[the remark this answers is not in the document]
Comment (Joseph-Edwards) [reply src/pbr.cpp:100]: You could call `product_inplace` and store the results in it? If that's worth anything?

**Comment by Joseph-Edwards** [comment]
> Can you maybe add the checked multiplication, and the other comparison operators? Shouldn't be much code

Is this simply the case of checking the degrees are the same? Or is there more to it?

File: src/pbr.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply src/pbr.cpp:100]: Ah, ok, sounds good

**Comment by james-d-mitchell** [comment]
> > Can you maybe add the checked multiplication, and the other comparison operators? Shouldn't be much code
> 
> Is this simply the case of checking the degrees are the same? Or is there more to it?

Just checking the degrees are equal and also validating the PBRs, that should do it.

**Pull request merged**